zenlink-protocol/src/swap/mod.rs

```diff
@@ -214,12 +215,21 @@ impl<T: Config> Pallet<T> {
 		path: &[T::AssetId],
 		recipient: &T::AccountId,
 	) -> DispatchResult {
-		let amounts = Self::get_amount_out_by_path(amount_in, path)?;
+		let mut new_amount_in = amount_in;
+		if path[0].is_native(T::SelfParaId::get()) {
+			// charge 0.5% going to pallet account for later distribution
+			let fee = amount_in.checked_div(200).unwrap_or_default();
+			new_amount_in = amount_in - fee;
```

Comment:
could do saturating here, but I don't see a scenario where overflow is possible so QED comment may work